File: BackEnd/models/studentModel.js
const Joi = require('joi');
const currentYear = new Date().getFullYear();

const studentSchema = Joi.object({
  nomor_pendaftaran: Joi.string().max(20),
  nama_lengkap: Joi.string().max(100).required(),
  nik: Joi.string().length(16).pattern(/^[0-9]+$/).required(),
  tempat_lahir: Joi.string().max(50).required(),
  tanggal_lahir: Joi.date().required().max(new Date()),
  jenis_kelamin: Joi.string().valid('L', 'P').required(),
  alamat: Joi.string().required(),
  provinsi: Joi.string().max(50).required(),
  kota_kabupaten: Joi.string().max(50).required(),
  kode_pos: Joi.string().max(10).required(),
  no_hp: Joi.string().max(15).pattern(/^[0-9]+$/).required(),
  email: Joi.string().email().max(100).required(),
  asal_sekolah: Joi.string().max(100).required(),
  tahun_lulus: Joi.number().integer().min(1900).max(currentYear).required()
});

const validateStudent = (data) => {
  return studentSchema.validate(data, { abortEarly: false });
};

module.exports = { validateStudent };

File: BackEnd/models/studentModel.go
package models

import (
	"time"

	"github.com/go-playground/validator/v10"
)

// Student represents the student model
type Student struct {
	UserID           string `json:"user_id" db:"user_id"`
	IDPendaftaran    string `json:"id_pendaftaran,omitempty" db:"id_pendaftaran"`
	NomorPendaftaran string `json:"nomor_pendaftaran,omitempty" db:"nomor_pendaftaran"`
	NamaLengkap      string `json:"nama_lengkap" binding:"required" validate:"required,max=100"`
	NIK              string `json:"nik" binding:"required" validate:"required,len=16,numeric"`
	TempatLahir      string `json:"tempat_lahir" binding:"required" validate:"required,max=50"`
	TanggalLahir     string `json:"tanggal_lahir" binding:"required" validate:"required,datetime=2006-01-02"`
	TanggalDaftar    string `json:"tanggal_daftar,omitempty" db:"tanggal_daftar"`
	JenisKelamin     string `json:"jenis_kelamin" binding:"required" validate:"required,oneof=L P"`
	Alamat           string `json:"alamat" binding:"required" validate:"required"`
	Provinsi         string `json:"provinsi" binding:"required" validate:"required,max=50"`
	KotaKabupaten    string `json:"kota_kabupaten" binding:"required" validate:"required,max=50"`
	KodePos          string `json:"kode_pos" binding:"required" validate:"required,max=10"`
	NoHP             string `json:"no_hp" binding:"required" validate:"required,max=15"`
	Email            string `json:"email" binding:"required" validate:"required,email,max=100"`
	AsalSekolah      string `json:"asal_sekolah" binding:"required" validate:"required,max=100"`
	TahunLulus       int    `json:"tahun_lulus" binding:"required" validate:"required,min=2000,max=2025"`
}

// CustomValidator is a custom validation type
type CustomValidator struct {
	validator *validator.Validate
}

// ValidateStudent validates the student data
func ValidateStudent(student *Student) error {
	validate := validator.New()

	// Register custom validation for checking date is not in future
	validate.RegisterValidation("ltefield", func(fl validator.FieldLevel) bool {
		if fl.Field().Interface().(time.Time).After(time.Now()) {
			return false
		}
		return true
	})

	return validate.Struct(student)
}
